tests/ship_log.cpp

```diff
@@ -152,15 +206,8 @@ BOOST_DATA_TEST_CASE(basic_prune_test, bdata::xrange(2) * bdata::xrange(2) * bda
       t.check_range_present(4, 7);
    });
 
-   //undo 6 & 7 and reapply 6
+   //undo 6 & 7 and reapply new 6 & 7
    t.add(6, payload_size, 'G', 'D');
-   t.check_n_bounce([&]() {
-      t.check_not_present(2);
-      t.check_not_present(3);
-      t.check_not_present(7);
-      t.check_range_present(4, 6);
-   });
-
```

Comment:
This test applies a fork that decreases the resulting head block number, and then closes the log and reopens it.  It was modified/removed because the new log implementation does not like this (to be specific, it fails on reopening the log file because the index contains more blocks than the log contains). Since this _is_ now valid in savanna the log will need to be improved.

For the simple single-log case this is quite a trivial fix: truncate the index as appropriate -- nothing in the ship connection thread could be using the index. For the partitioned log case things get a wee bit more tricky.

Reply:
There has been a fairly substantial modification of the logic in b98fe82 to ensure that the log is left in a sane state when a fork with a block number less than the previous head block is applied (this isn't strictly a problem when "closes the log and reopens it" as previously discussed). The change will require careful review.

For the single-log case it is as trivial as expected: truncate the index and then tweak some state of the log a little (arguably simpler now).

For partitioned logs, the approach is to "unrotate" the logs as needed: if the applied block occurs before the first block in the head log, the head log is removed and the newest retained log is restored as the head log. A new unit test has been enhanced to probe in this area more. I've also restored more of this section of the ship_log test that was removed previously.

#221 comes along for the ride in the improvement because now `random_access_file` and `state_history_log` are moveable (1876cc8 and 1eccd39)